### ISSUE TITLE
Steps towards Python 3.4 compatibility

### ISSUE THREAD

**Issue by ianozsvald** [coddingtonbear/python-myfitnesspal#19]
Hello, I'm a Python 3.4 user, your library seems to be for Python 2.7 only. I've done a little work to update a local copy, if others are on Python 3.4 then maybe I should try to build a PR?
I removed the `requirements.txt` (and the reference in `setup.py`) as the requirements are hardcoded to older versions of tools
Using Anaconda and `conda` I installed `requests` 2.7, `lxml` 3.4.4, `measurement` 1.5 (as you specified) and I didn't install `ordereddict`. 
References to `ordereddict` are replaced with `from collections import OrderedDict as ordereddict` as OrderedDict is built-in
A couple of unicode references are fixed

Python 2.7 has `OrderedDict` in collections, so the only reason to install `ordereddict` would be for e.g. Python 2.6 support (is that still supported?).

`mimic` installs but doesn't seem to have the methods you're expecting in the tests, so my conversion doesn't run `py.test` sadly.

Otherwise - cool! I've got data downloads now for diet analysis :-)

**Comment by coddingtonbear** [comment]
Hey there @ianozsvald!  It would be awesome to extend this library to have Python 3 support, and I'd love to help you out if you posted a PR, but there are just a couple things to make sure we're on the same page:

* We're trying to keep support for Python versions as low as 2.6 (see [our build results](https://travis-ci.org/coddingtonbear/python-myfitnesspal)).
* You really shouldn't remove the `requirements.txt`; as new libraries are released, they may introduce breaking changes (if they're following [semantic versioning](http://semver.org/) -- and most do -- that would be on the next "major" version), but do feel free to change the required versions to be ranged -- for example, replacing `requests==1.2.3` with `requests>=1.2.3,<=2.0.0` is totally fine!
* You can feel free to conditionally change the requirements list in your `setup.py` to add the third-party `ordereddict` library in situations where the package is being installed on Python 2.6 if you'd like (see [this example](https://github.com/ralphbean/taskw/blob/develop/setup.py#L40) from another library for an idea of how to accomplish that if you're unfamiliar).
* `gmcquillan-mimic` is a different package from `mimic` (although `gmcquillan-mimic` was originally `mimic`, somebody politely asked that package's maintainer if he could have the name for a different project).  That being said, I'm not married to continuing to use `gmcquillan-mimic` for mocking things out, and I'd totally be willing to help out in converting tests to using the now-standard [Mock](https://pypi.python.org/pypi/mock) library (which would, too, require conditionally adding it to requirements as it's built-in on Python 3.3+).

Thanks for offering to help out with adding Python 3 support, and do let me know what I can do to help.  If you have questions or want to chat, I've set up a [gitter channel](https://gitter.im/coddingtonbear/python-myfitnesspal) for this project, too.

Cheers, and have a nice weekend!

**Comment by ianozsvald** [comment]
Hi Adam. I'm curious about the support for Python 2.6 - it was deprecated around 2010 (when 2.7 was released) and 2.7 itself will be deprecated in 4.5 years (2020). Do you actually have users still running 2.6? I only ask because as co-org for the PyDataLondon group I'm continuously reminding folk to upgrade to Python 3, I rarely meet anyone using Python 2.6 for fun any more.
Re. removing `requirements.txt` - that was only to avoid forcing very-old libraries on my Python 3 env, I'd agree I wouldn't normally remove it!
Re. conditional `setup.py` flags for Python versions - hadn't seen that before - cheers!
Re. mock - I switched to `gmcquillan-mimic` and tried to port it to Python 3 but stumbled with unhashable types. Given that their last commit was 2013 I'm guessing that library isn't planning for Python 3 support.  Moving to Python 3's Mock is probably the right choice.
Re. time - I'm in the midst of organising a conference (PyDataLondon next weekend) so I shouldn't even be playing with this :-) I'll aim to setup a fork with a python3 branch with sane edits, I'm curious to hear your thoughts on Python 2.6 support (since Python 2.7 has `collections.OrderedDict`).

**Comment by coddingtonbear** [comment]
Hrmm, initially I was going to push hard on keeping Python 2.6 support since the effort involved is relatively minimal, but after doing a little investigation, even Django (which has a pretty rigorous policy) is no longer supporting versions earlier than Python 2.7.  OK; feel free to do things that eliminate support for Python 2.6.  Note that removing support for Python 2.6 would require that we do a major version bump in order to properly comply with semantic versioning.

Yeah, managing requirements files can be kind of an ordeal when supporting both Python 2.x and Python 3.x, but I'm afraid that defining compatible version ranges is the only way of making it possible for future installers of the library to be assured that the library will work properly in their environment.  Feel free to source from a different requirements file when installing on Python 3+ or otherwise alter the requirements list at runtime in your `setup.py`, but using a requirements file is the industry-standard method for handling python package dependencies and as tedious as it is to manage, I'm afraid it will have to remain.

About testing: yeah, I work with @gmcquillan day-to-day, and I'd be kind of astonished if he found the time to upgrade `gmcquillan-mimic` to Python 3, so we'll have to work together to move to a different mocking library (probably the now built-in Mock library).

So, since I've offered to help out where I can -- what I think we can do is, if you get started on a branch and push something up, let me know what the name of it is.  I'll then create my own branch and pull request, and pull your changes in as they appear.  How's that sound?

**Comment by ianozsvald** [comment]
Hey there. Here's a first port, compatible with python 2.7 (tested with nosetests) and python 3.4:
https://github.com/ianozsvald/python-myfitnesspal/tree/py34_compatibility

I'm using two Anaconda environments, I don't see anything that'll suggest the changes won't work outside of Anaconda.

The changes are very minor:
* setup.py unchanged (and untested in python 3.4)
* requirements.txt updated for >= version check
* decoded bytes from requests call in client.py, rather than decode in base.py (I think this is better, but I'm open to feedback)
* removed ordereddict and replaced with collections.OrderedDict
* I'm using lxml 3.4.4, requests 2.7, measurements 1.7.3
* added two PORTING_CHECK notes which you'll want to eyeball
* tiny bugfix (day.py)
* added `demo.py` script, you pass in username & password and it makes some calls as a trivial demo (to help with testing in Python 3.4 where the nosetests don't work as gmcquillan-mimic isn't supported)

I ran a coverage check (approx 80% per module), one of the missed bits is where I made the bugfix, probably we want a bit more coverage to be confident but I think my changes are pretty sensible. Let me know if you want a PR raised?

**Comment by ianozsvald** [comment]
I added a small fix in base.py using `six` this morning to handle unicode meals & notes (in python 2.7 this was barfing, now fixed). The solution is hacky but works. `six` is a dependency of one of the other modules (maybe `measurement`?) so it was already provided. Really the decoding from bytes (/str) should be handled at the point where the bytes come in, so we only have unicode on the inside of the app. This will mean chasing down their entry points and converting them, then `base.py`'s `__str__` decoding won't be necessary. 
I tried updating the `test_client` with a unicoded food name but `Day.get_as_dict` returns an ASCII version (at least, it stripped the unicode it seemed), so this might require some investigating.
For now this version runs on Python 2.7 & 3.4, reading a mix of plain-English and Unicode (Swedish!) food items and unicode notes (tested using `demo.py` in both Pythons), without barfing, and runs the existing Python 2.7 unit tests.
I'm happy to take a critique...
